ouroboros-consensus/src/Ouroboros/Consensus/Util/STM.hs

```diff
@@ -61,14 +62,16 @@ runWhenJust registry label getMaybeA action =
     void $ forkLinkedThread registry label $
       action =<< atomically (blockUntilJust getMaybeA)
 
-blockUntilJust :: MonadSTMTx stm => stm (Maybe a) -> stm a
+blockUntilJust :: MonadSTMTx stm tvar tmvar tqueue tbqueue
```

Comment:
Oof, this use site is now dissapointing.

Could we avoid the noise with a type alias like the following?

```
type MonadSTMTx stm = MonadSTMTx_ stm (TVar_ stm) (TMVar_ stm) ...
```

or does that trip the same quantified constraint limitation?

Reply:
Unfortunately this will not work, `MonadSTMTx` does not introduce any type families, they are defined at the level of `MonadSTM` and all are parametrised with `m` rather than `stm`.  Another choice is to use `MonadSTM m => Proxy m -> STM m (Maybe a) -> STM m a`

Reply:
I'll leave choice that to you; neither option is exciting :)

Reply:
I think this ought to be possible actually. `TVar_` _is_ introduced by `MonadSTMTx`:

```haskell
class ( Monad stm
      , Alternative stm
      , MonadPlus stm
      ) => MonadSTMTx stm where
  type TVar_ stm :: Type -> Type
```

`TVar` is merely a type alias:

```haskell
type TVar    m = TVar_    (STM m)
```

Reply:
This pr [removed `TVar_`](https://github.com/input-output-hk/ouroboros-network/pull/2977/files#diff-59c8b075c11f3c1bae4c6d47a75e36c6097679ae6d8a12879b1eeef2b3acdb66R111-L99) and replaced it with a [`TVar` type family](https://github.com/input-output-hk/ouroboros-network/pull/2977/files#diff-59c8b075c11f3c1bae4c6d47a75e36c6097679ae6d8a12879b1eeef2b3acdb66R189).

The crux of this pr is to substitute `TVar_` family (and friends) with a type variable to avoid `QuantifiedConstratint` restrictions.

Reply:
Ah, we missed that part of the diff -- were looking at master 🤦  Thanks for clarifying.

Reply:
I took a look, and I think I see a way forward. Specifically, a way that enables the `type` synonym in my first comment in this thread.

Please take a look at this commit: 96ebbdde1d682271b0ca88fcf716c0e5c84b76fd; `cabal v2-build all` succeeds on that commit.

I made no effort to pick good names. In fact, I picked at least one bad name -- I'm planning on you choosing the good names :)

Reply:
Also: I don't think the families have to be standalone. They could be associated with `MonadSTMTx`. The only reason I can think to not associate them with it is to avoid the oddity of that class having both arguments and type families that correspond to the same thing. Though, if we do go that route, the eventual comment on the class would explain that that is a symptom of the only workaround we were able to find...

Reply:
A few comments / questions:
* I think it would not it simplify the [`EqTVar`](https://github.com/input-output-hk/ouroboros-network/pull/2977/files#diff-59c8b075c11f3c1bae4c6d47a75e36c6097679ae6d8a12879b1eeef2b3acdb66R227) type alias.
* so we introduce another level of indirection, can GHC optimise it away?
* we only have two or three places where we use `MonadSTMTx`, is it worth it to optimise for this use case?